typescript/cli/src/utils/output.ts

```diff
@@ -0,0 +1,156 @@
+import chalk from 'chalk';
+
+import { CheckerViolation } from '@hyperlane-xyz/sdk';
+import { deepEquals, isNullish, isObject } from '@hyperlane-xyz/utils';
+
+interface ViolationOutput
+  extends Pick<CheckerViolation, 'actual' | 'expected'> {}
+
+export type ViolationDiff =
+  | {
+      [key: string]: ViolationOutput | ViolationDiff;
+    }
+  | ViolationDiff[];
+
+/**
+ * Takes 2 values and sorts them by the `type` field if they have it.
+ * This helps when comparing `IsmConfig` or `HookConfig` arrays
+ */
+export function sortByType(a: any, b: any): number {
+  if (a.type < b.type) {
+    return -1;
+  }
+
+  if (a.type > b.type) {
+    return 1;
+  }
+
+  return 0;
+}
+
+/**
+ * Merges 2 objects showing any difference in value for common fields.
+ */
+export function diffObjMerge(
+  actual: Record<string, any>,
+  expected: Record<string, any>,
+  max_depth = 10,
+): [ViolationDiff, boolean] {
+  if (max_depth === 0) {
+    throw new Error('diffObjMerge tried to go too deep');
+  }
+
+  let isDiff = false;
+  if (deepEquals(actual, expected)) {
+    return [actual, isDiff];
+  }
+
+  if (isObject(actual) && isObject(expected)) {
+    const ret: Record<string, ViolationDiff> = {};
+
+    const actualKeys = new Set(Object.keys(actual));
+    const expectedKeys = new Set(Object.keys(expected));
+    const allKeys = new Set([...actualKeys, ...expectedKeys]);
+    for (const key of allKeys.values()) {
+      if (actualKeys.has(key) && expectedKeys.has(key)) {
+        const [obj, diff] = diffObjMerge(
+          actual[key],
+          expected[key],
+          max_depth - 1,
+        );
+        ret[key] = obj;
+        isDiff ||= diff;
+      } else if (actualKeys.has(key) && !isNullish(actual[key])) {
+        ret[key] = {
+          actual: actual[key],
+          expected: '' as any,
+        };
+      } else if (!isNullish(expected[key])) {
+        ret[key] = {
+          actual: '' as any,
+          expected: expected[key],
+        };
+      }
+    }
+    return [ret, isDiff];
+  }
+
+  // Merge the elements of the array to see if there are any differences
+  if (
+    Array.isArray(actual) &&
+    Array.isArray(expected) &&
+    actual.length === expected.length
+  ) {
+    // Sorting because there might be cases where the arrays have the same elements but
+    // in different order.
```

Comment:
So if we have 2 arrays with the same elements but different ordering we should flag it as a mismatch then? I added the sort logic to avoid cases like this one

![image](https://github.com/user-attachments/assets/97bb9778-c562-4e54-8ffd-65c578f544f5)

which I think might seem weird to be reported as a mismatch when the config is technically correct but the only issue is the ordering

Reply:
fwiw we sort arrays as part of the config normalization step in the core/hook/ism modules before doing object comparisons

https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/adedb3db8e993805309f248bdb800486a8b69413/typescript/sdk/src/utils/ism.ts#L18
https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/adedb3db8e993805309f248bdb800486a8b69413/typescript/sdk/src/utils/ism.ts#L40-L41
https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/adedb3db8e993805309f248bdb800486a8b69413/typescript/sdk/src/ism/EvmIsmModule.ts#L148-L158